csharp/src/Google.Protobuf/Reflection/Dynamic/FieldSet.cs

```diff
@@ -0,0 +1,374 @@
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Checks if the fields dictionary contains the field.
+        /// </summary>
+        private bool HasField(FieldDescriptor field)
+        {
+            if (field.IsRepeated)
+            {
+                throw new ArgumentException("HasField() can only be called on non-repeated fields.");
+            }
+
+            return fields.ContainsKey(field);
+        }
+
+        /// <summary>
+        /// Returns the serialized size.
+        /// </summary>
+        /// <returns></returns>
+        internal int GetSerializedSize()
+        {
+            int size = 0;
+            foreach (KeyValuePair<FieldDescriptor, object> kvPair in fields)
+            {
+                size += ComputeSize(kvPair.Key, kvPair.Value);
+            }
+            return size;
+        }
+
+        /// <summary>
+        /// Computes the size of the field. The computation varies if the field is repeated.
+        /// </summary>
+        /// <param name="desc"></param>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private int ComputeSize(FieldDescriptor desc, object value)
+        {
+            if (desc.IsRepeated)
+            {
+                int dataSize = 0;
+                int tagSize = CodedOutputStream.ComputeTagSize(desc.FieldNumber);
+                foreach (object val in (IEnumerable) value)
+                {
+                    dataSize += ComputeElementSizeNoTag(desc.FieldType, val) + tagSize;
+                }
+                return dataSize;
+            }
+            else
+            {
+                return ComputeElementSize(desc.FieldType, desc.FieldNumber, value);
+            }
+
+        }
+
+        /// <summary>
+        /// Computes the total size of the filed, along with its tag.
+        /// </summary>
+        /// <param name="fieldType"></param>
+        /// <param name="fieldNumber"></param>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private int ComputeElementSize(FieldType fieldType, int fieldNumber, object value)
+        {
+            int tagSize = CodedOutputStream.ComputeTagSize(fieldNumber);
+            return tagSize + ComputeElementSizeNoTag(fieldType, value);
+        }
+
+        /// <summary>
+        /// Computes the size of the field without tag.
+        /// </summary>
+        /// <param name="fieldType"></param>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentException"></exception>
+        private int ComputeElementSizeNoTag(FieldType fieldType, object value)
+        {
+            switch (fieldType)
+            {
+                case FieldType.String:
+                    return CodedOutputStream.ComputeStringSize(value.ToString());
+                case FieldType.Bool:
+                    return CodedOutputStream.ComputeBoolSize(bool.Parse(value.ToString()));
```

Comment:
This pattern looks suspect across all cases of this switch. Why are you going ToString() and then re-parsing the value?